akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -521,6 +521,25 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
   def map[T](f: function.Function[Out, T]): javadsl.Flow[In, T, Mat] =
     new Flow(delegate.map(f.apply))
 
+
+  /**
+    * Filters and flattens, stream elements of type java.util.Optional and returns non empty elements
+    * as they pass through this processing step.
+    *
+    * '''Emits when''' the mapping function returns an element
+    *
+    * '''Backpressures when''' downstream backpressures
+    *
+    * '''Completes when''' upstream completes
+    *
+    * '''Cancels when''' downstream cancels
+    */
+  def flattenOptional[T](): javadsl.Flow[In, T, Mat] =
+    new Flow(
+      delegate.filter(e=>e.isInstanceOf[Optional[_]] && (e.asInstanceOf[Optional[_]]).isPresent)
+        .map(e=>e.asInstanceOf[Optional[T]].get())
```

Comment:
Implement it with collect instead of filter + map, also `T` _must_ be an optional so it should be something more clever here that translates to Java with a type bound giving a compile error if `T` is not an optional.

Reply:
I dont think its possible, to throw compile error if its not opitonal. Because flatten wont take any arguements. I think its not possible in java to throw compile error for function which doesnt take any arguement.

Reply:
It couldn't be an inline flow operator because of the Java type system limitations but I think it could be done if we moved it to a static method. Something like:

```
def flattenOptional[Out, In <: Optional[Out], Mat](src: Source[In, Mat]): Source[Out, Mat] 
```

I'm not 100% sure that is worth it, from an ergonomics/discoverability point of view, but I think allowing creating possibly invalid operators (no bound on T means I can call it with whatever type) is pretty bad.

Reply:
"Out, In <: Optional[Out]"   Out and In are class generic type name,
that we cant use it to restrict method for that class.
Else we need to specify type whenever we call which can be different one from class generic type.
Example,
import java.util.Optional

class Flow[In,Out] {
  def flattenOptional1[Out,In<:Optional[Out]]:Out={return 1.asInstanceOf[Out];}
}


import java.util.Optional;

public class FlowTest {
    public static void main(String[] args) {
        Flow<Integer,Optional<String>> test = new Flow<>();

        //wont throw error
        test.flattenOptional1();

        //throws error
        test.<String,Optional<Integer>>flattenOptional1();

        //wont throw error
        test.<Integer,Optional<Integer>>flattenOptional1();

    }
}

Reply:
I think the problem is that you cannot use a type parameter from the surrounding class as a bound for the method type parameters in Java, that is why I mentioned using a static method instead.